Refactor build_data

jbcden opened the pull request:
This is a first pass at addressing https://github.com/steveklabnik/rustdoc/issues/55.

jbcden commented:
@steveklabnik Here's a first pass at refactoring the `build_data` function. I'm not sure if this is what you were thinking but let me know your thoughts!

steveklabnik commented on src/lib.rs:
```diff
@@ -321,10 +317,5 @@ struct Crate {
-#[derive(Debug)]
-enum Item {
-    Module { name: String, docs: String },
+    metadata: Vec<Metadata>,
```
👍

steveklabnik commented on src/lib.rs:
```diff
@@ -190,21 +192,15 @@ impl DocData {
-    ) -> Result<HashMap<String, Item>> {
-        let mut data = HashMap::new();
-
+    fn build_data(config: &Config, root_id: analysis::Id, krate: &mut Crate) -> Result<()> {
```
So, this still ends up leaving this function and taking a `&mut Crate`; what do you think about possibly inlining this into the constructor? does it being its own function make sense still?

jbcden replied:
@steveklabnik I go back and forth on that. As I said before, the main reason passing the crate in felt weird to me was that that seemed to signal that this function's main focus should be on the crate. Now that it is all about the crate, I don't feel this is too weird, though the name should definitely be changed to reflect this new role too.

steveklabnik replied:
yeah let's just inline it for now 👍

mgattozzi commented on src/lib.rs:
```diff
@@ -213,14 +209,11 @@ impl DocData {
         for def in defs.iter() {
```
I think it might make more sense to turn this from `iter` to `into_iter` that way the clones below can be avoided.

mgattozzi replied:
The reason being is defs isn't used after this loop at all so consuming the values should be fine.

steveklabnik replied:
yeah i was certainly cargo-culting this code; getting rid of eventual clones would be great!

jbcden commented:
> yeah let's just inline it for now 👍

@steveklabnik I inlined it in f91ddad.

jbcden commented:
@mgattozzi Thanks for the advice! I'm still pretty new to Rust so I'm sure I miss a lot of stuff like that. I switched to `into_iter` and removed the clones in f91ddad and d27a90e.

mgattozzi commented:
Woo awesome! :D
